### PR TITLE
[feat(whisper)] Add recognize_whisper

### PR DESCRIPTION
**Pull request by joy-void-joy**
Solve https://github.com/Uberi/speech_recognition/issues/624 by adding recognize_whisper to Recognizer.

This works by writing in a tempfile, due to the format whisper asks for.

Usage example:

```python
import speech_recognition as sr

# obtain audio from the microphone
r = sr.Recognizer()
with sr.Microphone() as source:
    print("Say something!")
    audio = r.listen(source)

print("Got it, now to recognize it...")

try:
    print("Whisper thinks you said " + r.recognize_whisper(audio, language='english'))
except sr.UnknownValueError:
    print("Whisper could not understand audio")
except sr.RequestError as e:
    print("Whisper error; {0}".format(e))
```

### CODE REVIEW

**Comment by ftnext** [comment]
Thanks!
I'll check this later.

File: tests/test_recognition.py
Comment (ftnext) [inline tests/test_recognition.py:98]: `model="small"` is required.

✍️ When I specify `model="base"` (the default value), this test failed due to wrong recognition.

```
======================================================================
FAIL: test_whisper_chinese (test_recognition.TestRecognition)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/.../speech_recognition-pr/tests/test_recognition.py", line 98, in test_whisper_chinese
    self.assertEqual(r.recognize_whisper(audio, language="chinese", **self.WHISPER_CONFIG), u"砸自己的腳")
AssertionError: "�<|translate|> I'm sorry." != '砸自己的腳'
- �<|translate|> I'm sorry.
+ 砸自己的腳
```

File: examples/microphone_recognition.py
Comment (ftnext) [inline examples/microphone_recognition.py:90]: It works!🎉 Thanks.

```
$ python examples/microphone_recognition.py
Say something!
/.../speech_recognition-pr/venv/lib/python3.9/site-packages/whisper/transcribe.py:78: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")
Whisper thinks you said  Hello whisper
```

**Review by ftnext — CHANGES_REQUESTED** [review]
Thanks a lot for your great PR.
Whisper works with SpeechRecognition😃

I am very sorry for my too late review.
I would like to merge this once only the MUST comment are addressed.

@joy-void-joy Can you respond to that comment?
If it's difficult for you, that is no problem.
I'll fix the MUST comment and merge this PR this night (JST)

Let's discuss comments other than MUST after merge.

File: speech_recognition/__init__.py
Comment (ftnext) [inline speech_recognition/__init__.py:1689]: memo: `or` is short-circuit.

https://docs.python.org/3/reference/expressions.html#boolean-operations
>The expression x or y first evaluates x; if x is true, its value is returned; otherwise, y is evaluated and the resulting value is returned.

- When you passed not empty `dict` as `load_options`, load model
- When `load_options` is `None` or `{}` and the instance does not have `whisper_model` attribute, then load model
- When `load_options` is `None` or `{}` and the instance have `whisper_model` attribute but the name `model` does not included, then load model

**Comment by ftnext** [comment]
It seems unit tests failed because of not pip installing whisper.
I'll fix the unittest workflow file to install.

**Comment by ftnext** [comment]
>FileNotFoundError: [Errno 2] No such file or directory: 'ffmpeg'

https://github.com/Uberi/speech_recognition/actions/runs/3405145980/jobs/5662903599

It seems that `ffmpeg` are needed to install in the ubuntu-latest runner.
FYI: https://github.com/actions/runner-images/tree/main/images/linux

**Comment by joy-void-joy** [comment]
Thanks for the review, do you need anything helped with? I think the fp16 bug should be fixed in the new version of whisper, I'd rather we didn't specify fp16=False if a GPU is available as there is a significant slowdown on CPU

**Comment by ftnext** [comment]
Thanks for your reply.

>I'd rather we didn't specify fp16=False if a GPU is available as there is a significant slowdown on CPU

I agree.
I already merged #630 `fp16=torch.cuda.is_available()` and I believe the current implementation is similar to your idea.

If you have an idea to make it even slightly better, please send us a pull request.
Pull requests are always welcome!

**Pull request merged**
